### PR TITLE
fix: modulo op with negative zero divisor produces Nan

### LINKED ISSUE #521: Modulus can produce incorrect results in some cases
### Describe the bug

Example of compatibility issue found by fuzz testing:

```
SELECT c90, c1, c90 % c1 FROM test1 ORDER BY c90, c1;
```

Spark: `-21840,-0.0,NULL`
Comet: `-21840,-0.0,NaN`

### Steps to reproduce

_No response_

### Expected behavior

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by vaibhawvipul**
## Which issue does this PR close?

Closes #521  and #665.

## What changes are included in this PR?

Improves compatibility with apache spark.

## How are these changes tested?

Added a relevant test case.

### CODE REVIEW

**Review by comphead — COMMENTED** [review]
imho it has to be fixed on DataFusion side. I can see it returns NaN
```
> select 1 % -0.0;
+------------------------+
| Int64(1) % Float64(-0) |
+------------------------+
| NaN                    |
+------------------------+
```

PG, Trino fails on such query, 
DuckDB, Spark returns NULL 

I'll start a ticket in DF, if community supports to go with NULL, so no action needed for this ticket, if they decide to fail the query we might need some custom handling

**Comment by comphead** [comment]
Filed https://github.com/apache/datafusion/issues/11051 lets see

**Comment by vaibhawvipul** [comment]
> imho it has to be fixed on DataFusion side. I can see it returns NaN
> 
> ```
> > select 1 % -0.0;
> +------------------------+
> | Int64(1) % Float64(-0) |
> +------------------------+
> | NaN                    |
> +------------------------+
> ```
> 
> PG, Trino fails on such query, DuckDB, Spark returns NULL
> 
> I'll start a ticket in DF, if community supports to go with NULL, so no action needed for this ticket, if they decide to fail the query we might need some custom handling

Make sense. I will keep my PR as it then until we get some resolution on the datafusion ticket. 

Thank you @comphead

**Comment by kazuyukitanimura** [comment]
The direct reason why `NaN` is produced by dividing by `-0.0` is that `EqualTo` condition is failing in `nullIfWhenPrimitive()` of `QueryPlanSerde`

DataFusion changed their behavior several versions ago, to fail for dividing by zero. The DataFusion community recommendation to keep old behavior was to put if condition such as `nullIfWhenPrimitive()`.

I would say we should do something like
```
if (expression.dataType == DoubleType) {
    If(Or(EqualTo(expression, zero), EqualTo(expression, negZero)), Literal.create(null, expression.dataType), expression)
} else {
    If(EqualTo(expression, zero), Literal.create(null, expression.dataType), expression)
}
```
`If` is also supported natively, so it won't be too much overhead (at least that was the case last time I tried).

**Comment by vaibhawvipul** [comment]
> The direct reason why `NaN` is produced by dividing by `-0.0` is that `EqualTo` condition is failing in `nullIfWhenPrimitive()` of `QueryPlanSerde`
> 
> DataFusion changed their behavior several versions ago, to fail for dividing by zero. The DataFusion community recommendation to keep old behavior was to put if condition such as `nullIfWhenPrimitive()`.
> 
> I would say we should do something like
> 
> ```
> if (expression.dataType == DoubleType) {
>     If(Or(EqualTo(expression, zero), EqualTo(expression, negZero)), Literal.create(null, expression.dataType), expression)
> } else {
>     If(EqualTo(expression, zero), Literal.create(null, expression.dataType), expression)
> }
> ```
> 
> `If` is also supported natively, so it won't be too much overhead (at least that was the case last time I tried). I would hesitate to re-implement whole operator only for negative zero case. I feel we should reuse DataFusion as much as possible

makes a lot of sense. I have mad a fix as per your comments.

please review @kazuyukitanimura

**Comment by vaibhawvipul** [comment]
can someone please re-run the CI? It seems like a task failed because of network error. 

cc @kazuyukitanimura @comphead

File: spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala [hunk 1]
```diff
@@ -2240,7 +2240,15 @@ object QueryPlanSerde extends Logging with ShimQueryPlanSerde with CometExprShim
...
-    If(EqualTo(expression, zero), Literal.create(null, expression.dataType), expression)
+    val negZero = Literal.create(-0.0, DoubleType)
+    if (expression.dataType == DoubleType) {
+      If(
+        Or(EqualTo(expression, zero), EqualTo(expression, negZero)),
+        Literal.create(null, expression.dataType),
+        expression)
+    } else {
+      If(EqualTo(expression, zero), Literal.create(null, expression.dataType), expression)
+    }
```
Comment (kazuyukitanimura) [inline spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:2251]: I gave another thought, the larger issue is that the following test fails
```
      withParquetTable(Seq((0.0, -0.0)), "tbl") {
        checkSparkAnswerAndOperator("SELECT _1 == _2 FROM tbl")
      }
```
So I think we need to fix EqualTo itself. I opened https://github.com/apache/datafusion/issues/11108

For short term, rather than fixing `nullIfWhenPrimitive`, we can consider fixing `EqualTo` using a similar technique. I.e.
```
 case EqualTo(left, right) =>
    val leftExpr = exprToProtoInternal( If(EqualTo(left, negZero), zero, left), inputs)
    val rightExpr = exprToProtoInternal( If(EqualTo(right, negZero), zero, right), inputs)
```
The above code creates infinity recursive loop, so needs some if-branches to prevent it...

Comment (vaibhawvipul) [reply spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala]: for now i have implemented in `nullIfWhenPrimitive`. 

I will explore now how to fix `EqualTo` and subsequently fix this issue.

Comment (kazuyukitanimura) [reply spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala]: Once `EqualTo` is fixed, we do not need `nullIfWhenPrimitive` workaround

Comment (vaibhawvipul) [reply spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala]: Hey @kazuyukitanimura , i tried a lot but I was not able to come up with a solution here. Do you have any pointers for me like how to do this in equalto?

Comment (vaibhawvipul) [reply spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala]: @kazuyukitanimura I take it back, ignore the above comment, I have made changes in EqualTo and now it behaves as expected.

i have also added a test case.

**Comment by vaibhawvipul** [comment]
@kazuyukitanimura ready for review

File: spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala [hunk 2]
```diff
@@ -856,6 +856,45 @@ class CometExpressionSuite extends CometTestBase with AdaptiveSparkPlanHelper {
...
+        (-0.0, 1.0),
+        (1.0, -0.0),
+        (1.0, -1.0),
+        (-1.0, 1.0),
+        (-1.0, -0.0),
+        (-1.0, -1.0),
+        (-1.0, 0.0),
+        (0.0, -1.0)),
+      "t") {
+      checkSparkAnswerAndOperator("SELECT _1 == _2 FROM t")
```
Comment (kazuyukitanimura) [inline spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala:879]: We may need to test `!=` and `<=>`as well, but can be a separate PR

File: spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala [hunk 3]
```diff
@@ -855,19 +855,57 @@ object QueryPlanSerde extends Logging with ShimQueryPlanSerde with CometExprShim
           None
 
         case EqualTo(left, right) =>
-          val leftExpr = exprToProtoInternal(left, inputs)
-          val rightExpr = exprToProtoInternal(right, inputs)
+          val zero = Literal.default(left.dataType)
+          val negZero = UnaryMinus(zero)
```
Comment (kazuyukitanimura) [inline spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:859]: RIght type may be different from left type. So we need Right version of `zero` and `negZero`?

For `negZero` you can do `Literal(-0.0)` and `Literal(-0.0f)` instead of UnaryMinus

File: spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala [hunk 4]
```diff
@@ -855,19 +855,57 @@ object QueryPlanSerde extends Logging with ShimQueryPlanSerde with CometExprShim
...
+              .newBuilder()
+              .setEq(
+                ExprOuterClass.Equal
+                  .newBuilder()
+                  .setLeft(leftExpr)
+                  .setRight(rightExpr))
+              .build()
+          }
+
+          if (left.dataType == DoubleType || left.dataType == FloatType) {
```
Comment (kazuyukitanimura) [inline spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:872]: What about `right.dataType`?

File: spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala [hunk 5]
```diff
@@ -855,19 +855,57 @@ object QueryPlanSerde extends Logging with ShimQueryPlanSerde with CometExprShim
...
+            // make sure left or right is not null
+            (left, right) match {
+              case (`negZero`, _) =>
```
Comment (kazuyukitanimura) [inline spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:875]: `case (negZero, negZero)` may fall into this case as well? If so only left side gets `Abs()`?

File: spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala [hunk 6]
```diff
@@ -855,19 +855,57 @@ object QueryPlanSerde extends Logging with ShimQueryPlanSerde with CometExprShim
...
+                return Some(
+                  buildEqualExpr(
+                    exprToProtoInternal(Abs(left).child, inputs).getOrElse(return None),
```
Comment (kazuyukitanimura) [inline spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:878]: It is better not to return `None` directly, because we would like to call `withInfo(expr, left, right)` for better logging

File: spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala [hunk 7]
```diff
@@ -855,19 +855,57 @@ object QueryPlanSerde extends Logging with ShimQueryPlanSerde with CometExprShim
...
-                .setEq(builder)
-                .build())
+          val (leftExpr, rightExpr) =
+            if (left.dataType == DoubleType || left.dataType == FloatType) {
+              (
+                exprToProtoInternal(If(EqualTo(left, negZero), zero, left), inputs),
+                exprToProtoInternal(If(EqualTo(right, negZero), zero, right), inputs))
+            } else {
+              (exprToProtoInternal(left, inputs), exprToProtoInternal(right, inputs))
+            }
```
Comment (kazuyukitanimura) [inline spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:905]: We should separate out left and right. I.e.
```
val leftExpr = if (left.dataType == DoubleType || left.dataType == FloatType) {
    exprToProtoInternal(If(EqualTo(left, negZero), zero, left), inputs)
} else {
    exprToProtoInternal(left, inputs)
}
val rightExpr = ...
```

File: spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala [hunk 8]
```diff
@@ -855,19 +855,57 @@ object QueryPlanSerde extends Logging with ShimQueryPlanSerde with CometExprShim
...
-          val leftExpr = exprToProtoInternal(left, inputs)
-          val rightExpr = exprToProtoInternal(right, inputs)
+          val zero = Literal.default(left.dataType)
+          val negZero = UnaryMinus(zero)
 
-          if (leftExpr.isDefined && rightExpr.isDefined) {
-            val builder = ExprOuterClass.Equal.newBuilder()
-            builder.setLeft(leftExpr.get)
-            builder.setRight(rightExpr.get)
+          def buildEqualExpr(leftExpr: Expr, rightExpr: Expr): ExprOuterClass.Expr = {
```
Comment (kazuyukitanimura) [inline spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:861]: Could you add a comment to mention this is a workaround until https://github.com/apache/datafusion/issues/11108 is resolved?

File: spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala (hunk 3 above)
Comment (vaibhawvipul) [reply spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala]: resolved.

File: spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala (hunk 2 above)
Comment (vaibhawvipul) [reply spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala:879]: sure, I will create a different PR for it.

**Comment by vaibhawvipul** [comment]
@kazuyukitanimura ready for review.

**Comment by vaibhawvipul** [comment]
@kazuyukitanimura could you please review once? CI passed.

**Comment by kazuyukitanimura** [comment]
@vaibhawvipul Sorry, I got busy with something else. I may be unable to review or work on OSS a week or two.

**Comment by vaibhawvipul** [comment]
@andygrove @viirya can i get a review please?

File: spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala [hunk 9]
```diff
@@ -982,23 +981,80 @@ object QueryPlanSerde extends Logging with ShimQueryPlanSerde with CometExprShim
...
+                  exprToProtoInternal(Abs(right).child, inputs))
+              case _ =>
+                val doubleNan = Literal(Double.NaN, DoubleType)
+                val floatNan = Literal(Float.NaN, FloatType)
+
+                if ((left.nullable && !right.nullable) &&
+                  (left != negZeroLeft && right != negZeroRight) &&
+                  (left != leftZero && right != rightZero) &&
+                  (left != doubleNan && right != doubleNan) &&
+                  (left != floatNan && right != floatNan)) {
```
Comment (kazuyukitanimura) [inline spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:1034]: Hmm do you mind adding a comment explaining what this condition is trying to check?

File: spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala [hunk 10]
```diff
@@ -982,23 +981,80 @@ object QueryPlanSerde extends Logging with ShimQueryPlanSerde with CometExprShim
...
+            }
+          }
 
-          if (leftExpr.isDefined && rightExpr.isDefined) {
-            val builder = ExprOuterClass.Equal.newBuilder()
-            builder.setLeft(leftExpr.get)
-            builder.setRight(rightExpr.get)
+          val leftExpr =
+            if (left.dataType == DoubleType || left.dataType == FloatType) {
+              exprToProtoInternal(If(EqualTo(left, negZeroLeft), leftZero, left), inputs)
```
Comment (kazuyukitanimura) [inline spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:1046]: Hmm `exprToProtoInternal` recursively gets executed  and `EqualTo(left, negZeroLeft)` will match the above case-match and gets converted at
```
              case (_, `negZeroRight`) =>
                 return buildEqualExpr(
                   exprToProtoInternal(left, inputs),
                   exprToProtoInternal(Abs(right).child, inputs))
```
??

Comment (vaibhawvipul) [reply spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:1049]: Yes

File: spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala (hunk 9 above)
Comment (vaibhawvipul) [reply spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:1039]: added comments to explain this. @kazuyukitanimura

**Comment by vaibhawvipul** [comment]
@andygrove / @kazuyukitanimura can we please get CI triggered? and also a review?

File: spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala [hunk 11]
```diff
@@ -985,23 +984,82 @@ object QueryPlanSerde extends Logging with ShimQueryPlanSerde with CometExprShim
...
+                  exprToProtoInternal(Abs(left).child, inputs),
+                  exprToProtoInternal(Abs(right).child, inputs))
+              case (`negZeroLeft`, _) =>
+                return buildEqualExpr(
+                  exprToProtoInternal(Abs(left).child, inputs),
+                  exprToProtoInternal(right, inputs))
+              case (_, `negZeroRight`) =>
+                return buildEqualExpr(
+                  exprToProtoInternal(left, inputs),
+                  exprToProtoInternal(Abs(right).child, inputs))
```
Comment (kazuyukitanimura) [inline spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:1028]: Do you mind explaining why we need `Abs(right).child`? I thought  `Abs(right).child == right`

File: spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala [hunk 12]
```diff
@@ -985,23 +984,82 @@ object QueryPlanSerde extends Logging with ShimQueryPlanSerde with CometExprShim
...
+              case _ =>
+                val doubleNan = Literal(Double.NaN, DoubleType)
+                val floatNan = Literal(Float.NaN, FloatType)
+
+                // Ensure neither left nor right is -0.0 or 0.0 or NaN
+                // also return none if one side is nullable and the other is not
+                if ((left.nullable && !right.nullable) &&
```
Comment (kazuyukitanimura) [inline spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:1035]: What is the reason of the `left.nullable && !right.nullable` requirement?

File: spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala (hunk 10 above)
Comment (kazuyukitanimura) [reply spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:1051]: Hmmm still trying to understand how this recursion works...

**Comment by kazuyukitanimura** [comment]
The main issue was fixed by https://github.com/apache/datafusion-comet/pull/953
The test cases from this PR would be still helpful

**Comment by kazuyukitanimura** [comment]
Please never mind my comment above, I will post a workaroud PR

**Comment by comphead** [comment]
@kazuyukitanimura can we close it?

**Comment by kazuyukitanimura** [comment]
Yes @comphead @vaibhawvipul
